### PR TITLE
Beginnings of weapons refactor.

### PR DESCRIPTION
**Pull request by Necromunger**
weapons system remake

### CODE REVIEW

File: Assets/scripts/PlayGroups/PlayerScript.cs [hunk 1]
```diff
@@ -89,9 +89,10 @@ public class PlayerScript: NetworkBehaviour {
         }
 
         public bool IsInReach(Transform transform) {
-            if(pickUpCoolDown)
-                return false;
-            StartCoroutine(PickUpCooldown());
+            //if(pickUpCoolDown)
+            //    return false;
+            //StartCoroutine(PickUpCooldown());
```
Comment (kaffo) [inline Assets/scripts/PlayGroups/PlayerScript.cs]: Remove comment :)

File: Assets/scripts/Weapons/Weapon.cs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+					return;
+				} 
+				//if we are out of ammo for this weapon play an empty sound
+				else if (Projectile != null && CurrentMagazine.ammoRemains <= 0 && FireCountDown <= 0) {
+					InAutomaticAction = false;
+					PlayEmptySFX();
+					return;
+				}
+				//if we are out of ammo for this weapon play an empty sound - machine gun fire ending
+				else if (Projectile != null && CurrentMagazine.ammoRemains <= 0 && FireCountDown <= 0 && InAutomaticAction) {
```
Comment (kaffo) [inline Assets/scripts/Weapons/Weapon.cs:159]: Won't the if before this always be true first and we'll never go down this path?

File: Assets/scripts/Weapons/Weapon.cs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+		}
+
+		void AttemptToFireWeapon()
+		{
+			//ignore if we are hovering over UI
+			if (UnityEngine.EventSystems.EventSystem.current.IsPointerOverGameObject())
+				return;
+
+			//if the hand with the weapon in it is selected
+			if (UIManager.Hands.CurrentSlot == UIManager.Hands.LeftSlot || UIManager.Hands.CurrentSlot == UIManager.Hands.RightSlot) {
```
Comment (kaffo) [inline Assets/scripts/Weapons/Weapon.cs:145]: If there's a check for the FireCountDown on every path except the first one, can we wrap the rest of the paths in one 
`if (FireCoutnDown <= 0`
Means it doesn't bother checking the other clauses every time

File: Assets/scripts/Weapons/Weapon.cs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+				else {
+					//if we have a projectile to shoot, we have ammo and we are not waiting to be allowed to shoot again, Fire!
+					if (Projectile != null && CurrentMagazine.ammoRemains > 0 && FireCountDown <= 0) {
+						//Add too the cooldown timer to being allowed to shoot again
+						FireCountDown += 1.0 / FireRate;
+						//fire a single round if its a semi or automatic weapon
+						if (WeaponType == WeaponType.SemiAutomatic || WeaponType == WeaponType.FullyAutomatic) {
+							Vector2 dir = (Camera.main.ScreenToWorldPoint (Input.mousePosition) - PlayerManager.LocalPlayer.transform.position).normalized;
+							PlayerManager.LocalPlayerScript.weaponNetworkActions.CmdShootBullet(dir, Projectile.name);
+							CurrentMagazine.ammoRemains--; //TODO: remove more bullets if burst
```
Comment (kaffo) [inline Assets/scripts/Weapons/Weapon.cs:173]: I know it's on the TODO, but burst isn't even in this path (yet) anyway :)

**Review by kaffo — APPROVED** [review]
Check these out and change them if you want. But seems good otherwise!

**Comment by DooblyNoobly** [comment]
Resolved conflicts, and merged

File: Assets/scripts/Weapons/Weapon.cs (hunk 2 above)
Comment (Necromunger) [reply Assets/scripts/Weapons/Weapon.cs:159]: those if's should be swapped around
so 
`else if (Projectile != null && CurrentMagazine.ammoRemains <= 0 && FireCountDown <= 0 && InAutomaticAction) {`
should come first

**Pull request merged**